Basic EME

dmlap opened the pull request:
Listen for EME v0.1b events on the video element and handle a very basic authorization flow. The license server communication is expected to be handled by EME implementors. An example of what that would look like is provided in example.html.

The decryption code emulates the API of the current version of the EME spec. When the updated spec is released into the stable version of Chrome, we can rip out the emulation and just use the native APIs directly.

dmlap committed:
Add needkey listeners
Listen for "needkey" events on the video element and request keys.

dmlap committed:
Working example of EME with DASH/CENC
Create and hook up the decrypter module when the DASH source handler is invoked. Emulate the latest version of the EME spec by implementing it against v0.1b, the version that is available in Chrome 40. Implement default content protection handling that would allow a downstream developer to listen for "mediakeymessage" on the player and then supply a license to play protected media. This mechanism is much less flexible than the full set of EME APIs but is also a lot less coding to integrate against. Generalize segment template duration calculation to return the maximum duration of children segment elements if a duration is not available at the template level. Ensure relative MPD URLs are concatenated with forward-slashes joining path parts. Generalize media set lookups to search representation children of adapation set elements in addition to the adapation set itself. Switch the MPD type of the EME fixture manifest to "static" to avoid the incomplete dynamic manifest handling logic.

gkatsev commented on src/js/dash/segments/getSegmentListForRepresentation.js:
```diff
@@ -131,7 +144,7 @@ createSegmentListFromTemplate = function(representationXml) {
 
                 initializationRelativeUrl = segmentTemplate.replaceTokenForTemplate(initializationRelativeUrl, 'Bandwidth', representationXml.getBandwidth());
-                return baseUrl + initializationRelativeUrl;
+                return [baseUrl, initializationRelativeUrl].join('/');
```
is this known to produce good urls? Would we end up with extra slashes? Do we care about that?
Given that we use browserify, we could always require the `url` module and [resolve it](https://iojs.org/api/url.html#url_url_resolve_from_to).

dmlap replied:
I'm not sure :) I did a couple quick fixes to get the demo working but a more comprehensive review of a lot of the MPD spec and the handling in this project should be on our TODOs.

gkatsev commented on test/decrypter.test.js:
```diff
@@ -1,5 +1,153 @@
+});
+
+Q.test('throws early if used in a browser without EME support', function() {
+  try {
```
QUnit has a [`throws`](http://api.qunitjs.com/throws/) method that can be used for this.

dmlap replied:
Oh yeah. I always forget about that one.

gkatsev commented on test/fixtures/car.mpd:
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<MPD xmlns="urn:mpeg:DASH:schema:MPD:2011" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xlink="http://www.w3.org/1999/xlink" xsi:schemaLocation="urn:mpeg:DASH:schema:MPD:2011 DASH-MPD.xsd" minBufferTime="PT2S" type="dynamic" profiles="urn:mpeg:dash:profile:isoff-live:2011" availabilityStartTime="2014-09-19T00:54:13" minimumUpdatePeriod="PT5S" timeShiftBufferDepth="PT1800S">
+<MPD xmlns="urn:mpeg:DASH:schema:MPD:2011" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xlink="http://www.w3.org/1999/xlink" xsi:schemaLocation="urn:mpeg:DASH:schema:MPD:2011 DASH-MPD.xsd" minBufferTime="PT2S" type="static" profiles="urn:mpeg:dash:profile:isoff-live:2011" availabilityStartTime="2014-09-19T00:54:13" minimumUpdatePeriod="PT5S" timeShiftBufferDepth="PT1800S">
```
what's the difference between the `static` and `dynamic` `type`s?

dmlap replied:
`dynamic` in this code triggers live-style playback. Given this is pointing to a VOD video, I think that's probably an unsafe simplification but I'm not familiar enough with MPD to know for sure.

dmlap committed:
Use Q.throws
Take advantage of the QUnit helper method for checking when a block throws an exception.

gkatsev commented:
LGTM
